### ISSUE TITLE
textDocument/documentSymbol not working with hierarchical outline

### ISSUE THREAD

**Issue by WebFreak001** [Microsoft/vscode-languageserver-node#406]
I have implemented hierarchical outlining in my language server and vscode calls it on ctrl-shift-o, but it is never displaying it, only "No symbols found". Log:

```
2018-08-26T15:23:40.449:app.d:processRequest:77 Calling provideDocumentSymbols
[Trace - 3:23:40 PM] Received response 'textDocument/documentSymbol - (24)' in 17ms.
Result: [
    {
        "children": [],
        "deprecated": false,
        "detail": null,
        "kind": 12,
        "name": "main",
        "range": {
            "end": {
                "character": 0,
                "line": 19
            },
            "start": {
                "character": 0,
                "line": 9
            }
        },
        "selectionRange": {
            "end": {
                "character": 0,
                "line": 9
            },
            "start": {
                "character": 0,
                "line": 9
            }
        }
    }
]
```

The outline panel is also not showing anything.

**Comment by rcjsuen** [comment]
Implement your own custom `Middleware` and try to intercept the call to see what's going on.

https://github.com/Microsoft/vscode-languageserver-node/blob/5b8c5584fa46dfbb2b35115014007c7ed332dc61/client/src/client.ts#L403

**Comment by WebFreak001** [comment]
so it is receiving the correct stuff but it parses it as SymbolInformation instead of a DocumentSymbol. Could it be that null is not treated as string by that checker function?

https://github.com/Microsoft/vscode-languageserver-node/blob/5b8c5584fa46dfbb2b35115014007c7ed332dc61/client/src/client.ts#L1607

**Comment by dbaeumer** [comment]
This is true. Right now optional values must be `undefined` and not `null`. This is actually consistent in the whole `vscode-languageserver` project. Slightly related to https://github.com/Microsoft/language-server-protocol/issues/355.

@aeschli & @rcjsuen any opinion on this on allowing `null` for optional values. I don't see a reason for not allowing this although we need to think whether this would break the specification. 

@WebFreak001 which language is your server implemented in?

**Comment by WebFreak001** [comment]
my server is implemented in D, where empty strings and arrays are equal to a null slice (0 length, null ptr), which is serialized by the json serializer I use by default to `null`. I have now added a custom serialization policy for the structs in question but I think it would be nice if the parser would be a bit more leniant on allowing null on values, especially because when using fixed data types in some languages it might not really be possible to define policies like ignoring unchanged values without a lot of hassle.

**Comment by rcjsuen** [comment]
I'm concerned about backwards compatibility but as the LSP (hopefully) continues to growth, we're going to start seeing more of these minute differences in JSON serialization and language types.

There's never going to a better time to (possibly) break things besides now. Tomorrow there will be one more LSP server or one more LSP client so let's just do it.

**Comment by dbaeumer** [comment]
I read again through both issues and the only valid way forward would be to clearly state which of the optional properties can carry `null`. And even if we do this it will be a breaking change. So we would need to guard this using client and server capabilities. The question is if this is really worth the swirl.

**Comment by WebFreak001** [comment]
I have migrated to a custom serializer and marked the problematic areas as ignore-null in my code, so as long as we don't make it more strict than it is now I'm personally fine with staying how it is.

The `?` in the specification does already mark the optional ones, though it doesn't necessarily specify what a missing value means in all cases, when it is probably null in most implementations.
